File: backend/app/schemas/overview.py
"""Overview schemas.

One shape serves both audiences. A client token gets a single-element `tenants`
list and `scope: "tenant"`; a staff fleet token gets one row per client and the
platform-level warnings. Building two overviews would guarantee they drift.
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class TenantOverview(BaseModel):
    tenant_id: uuid.UUID
    name: str
    slug: str
    colour: str | None = None
    status: str

    open_incidents: int = 0
    new_incidents: int = 0
    unassigned_incidents: int = 0
    # Severity 13+ on the Wazuh ramp: the ones that get someone out of bed.
    critical_open: int = 0
    response_breached: int = 0
    resolution_breached: int = 0
    # Open, unbreached, and inside the last hour of its response deadline.
    at_risk: int = 0
    # Clock stopped awaiting the client. Counted separately because these cannot
    # breach, and an analyst reading a queue needs to know why.
    awaiting_client: int = 0

    alerts_24h: int = 0
    last_alert_at: datetime | None = None
    # True when this client has never delivered an alert, or has gone quiet for
    # longer than a day. Almost always an integration problem, not a quiet week.
    silent: bool = False

    agents_total: int = 0
    agents_active: int = 0
    agents_disconnected: int = 0
    last_sync_at: datetime | None = None
    last_sync_error: str | None = None
    has_connection: bool = False
    has_sla: bool = False


class MisgroupedAgent(BaseModel):
    """The one cross-tenant leak ingest cannot catch - see `wazuh/sync.py`."""

    base_url: str
    agent_id: str
    agent_name: str | None = None
    tenant_slugs: list[str] = Field(default_factory=list)


class Overview(BaseModel):
    scope: str
    generated_at: datetime
    tenants: list[TenantOverview] = Field(default_factory=list)

    open_incidents: int = 0
    critical_open: int = 0
    response_breached: int = 0
    at_risk: int = 0
    alerts_24h: int = 0

    # Staff only. Empty for a client token, which must never learn that another
    # tenant exists at all.
    misgrouped_agents: list[MisgroupedAgent] = Field(default_factory=list)
    silent_tenants: list[str] = Field(default_factory=list)

File: frontend/src/api/overview.ts
import { api } from "./client";

export interface TenantOverview {
  tenant_id: string;
  name: string;
  slug: string;
  colour: string | null;
  status: string;

  open_incidents: number;
  new_incidents: number;
  unassigned_incidents: number;
  critical_open: number;
  response_breached: number;
  resolution_breached: number;
  at_risk: number;
  /** Clock stopped awaiting the client. These cannot breach. */
  awaiting_client: number;

  alerts_24h: number;
  last_alert_at: string | null;
  silent: boolean;

  agents_total: number;
  agents_active: number;
  agents_disconnected: number;
  last_sync_at: string | null;
  last_sync_error: string | null;
  has_connection: boolean;
  has_sla: boolean;
}

/** The one cross-tenant leak ingest cannot catch: the same agent id cached
 *  under two clients on the same manager means their groups overlap. */
export interface MisgroupedAgent {
  base_url: string;
  agent_id: string;
  agent_name: string | null;
  tenant_slugs: string[];
}

export interface Overview {
  scope: "tenant" | "fleet";
  generated_at: string;
  tenants: TenantOverview[];
  open_incidents: number;
  critical_open: number;
  response_breached: number;
  at_risk: number;
  alerts_24h: number;
  misgrouped_agents: MisgroupedAgent[];
  silent_tenants: string[];
}

export const getOverview = () => api<Overview>("/overview");

export interface TimeBucket {
  at: string;
  incidents: number;
  alerts: number;
  critical: number;
}

export interface SeveritySlice {
  label: string;
  severity_min: number;
  count: number;
}

export interface StatusSlice {
  status: string;
  count: number;
}

export interface OverviewTrend {
  since: string;
  /** 1 for the hourly view, 24 for the daily one. The chart labels its axis
   *  from this rather than guessing from the point spacing. */
  bucket_hours: number;
  buckets: TimeBucket[];
  by_severity: SeveritySlice[];
  by_status: StatusSlice[];
}

export const getTrend = (days = 7) =>
  api<OverviewTrend>(`/overview/trend?days=${days}`);
